hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -38,32 +80,151 @@ impl TraceDecoder for YkPTTraceDecoder {
         let itr = YkPTBlockIterator {
             errored: false,
             parser: PacketParser::new(trace.bytes()),
-            cur_loc: ObjLoc::OtherObjOrUnknown,
+            cur_loc: ObjLoc::OtherObjOrUnknown(None),
             tnts: VecDeque::new(),
+            comprets: CompressedReturns::new(),
             pge: false,
         };
         Box::new(itr)
     }
 }
 
+/// The virtual address ranges of segments that we may need to disassemble.
+static CODE_SEGS: LazyLock<CodeSegs> = LazyLock::new(|| {
+    let mut segs = Vec::new();
+    for obj in phdrs::objects() {
+        let obj_base = obj.addr();
+        for hdr in obj.iter_phdrs() {
+            if (hdr.flags() & libc::PF_W) == 0 {
+                let vaddr = usize::try_from(obj_base + hdr.vaddr()).unwrap();
+                let memsz = usize::try_from(hdr.memsz()).unwrap();
+                let key = vaddr..(vaddr + memsz);
+                segs.push((key.clone(), ()));
+            }
+        }
+    }
+    let tree = segs.into_iter().collect::<IntervalTree<usize, ()>>();
+    CodeSegs { tree }
+});
+
+const PT_MAX_COMPRETS: usize = 64;
+
+/// A data structure providing convenient access to virtual address ranges and memory slices for
+/// segments.
+///
+/// FIXME: For now this assumes that no dlopen()/dlclose() is happening.
+struct CodeSegs {
+    tree: IntervalTree<usize, ()>,
+}
+
+impl CodeSegs {
+    /// Obtain the virtual address range and a slice of memory for the segment containing the
+    /// specified virtual address.
+    fn seg<'s: 'a, 'a>(&'s self, vaddr: usize) -> Segment<'a> {
+        let mut hits = self.tree.query(vaddr..(vaddr + 1));
```

Comment:
Is this a `for` loop in disguise? i.e. can I do `for x in self.tree.query(vaddr..(vaddr + 1))`?

Reply:
It's like this because generally speaking an interval tree query can return more than one result, but in our scenario that is impossible because our query is a point interval (a range of one element, which cannot span multiple segments). Due to this there should be exactly zero or one result.

Reply:
But after this line is a `while` loop that pumps `hits`?

Reply:
I don't think so.

We just want to make sure that we consume a sole result from the query. In debug mode we call `next()` twice to ensure there isn't another element. I'm not sure how a while loop would factor into this.